mount/mount_darwin.go

```diff
@@ -0,0 +1,85 @@
+//go:build darwin
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package mount
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	exec "golang.org/x/sys/execabs"
+)
+
+// Mount to the provided target.
+func (m *Mount) mount(target string) error {
+	var commandName string
+	if m.Type == "bind" {
+		// macOS doesn't natively support bindfs/nullfs
+		// The way to emulate it is via FUSE fs named "bindfs"
+		commandName = "bindfs"
+	} else {
+		commandName = fmt.Sprintf("mount_%s", m.Type)
+	}
+
+	var args []string
+	for _, option := range m.Options {
+		if option == "rbind" {
+			// On one side, rbind is not supported by macOS mounting tools
+			// On the other, bindfs works as if rbind is enabled anyway
+			continue
+		}
+
+		args = append(args, "-o", option)
+	}
+	args = append(args, m.Source, target)
+
+	// We can't use exec.Command here because bindfs forks "daemon" process that keeps stderr/stdout open and hangs Command.Wait
+	// Instead, the idea of the following code is:
+	// 1. We start the process and wait for its exit code
+	// 2. If it is successful, we cancel stdout/stderr reading
+	// 3. If it is not successful, we read the whole stdout/stderr
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	cmd := exec.CommandContext(ctx, "bindfs", args...)
+	var buf bytes.Buffer
+	cmd.Stdout = &buf
+	cmd.Stderr = &buf
```

Comment:
Maybe give cmd.StdoutPipe/ErrPipe a try, as it assigns os.File's to stdio. From the os/exec blurb above `Wait` 
```
// If any of c.Stdin, c.Stdout or c.Stderr are not an *os.File, Wait also waits
// for the respective I/O loop copying to or from the process to complete.
```

if it daemonizes it's likely stuck here https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/os/exec/exec.go;l=908

Reply:
> if it daemonizes it's likely stuck here

yes, exactly.

Reply:
I need some time for this, will experiment later today. Hopefully, `cmd.StdoutPipe/ErrPipe` will allow to simplify the code (yes, I agree that it is awful).

Reply:
ack, just gimme a ping

Reply:
Good news! I managed to throw away all this stuff. It turns out, that unclosed streams issue only [affects fuse-t](https://github.com/macos-fuse-t/fuse-t/issues/30) and is very likely a bug that I do not to workaround given that there are even [more bugs](https://github.com/macos-fuse-t/fuse-t/issues/29). Now `mount` function on Darwin is very simple and nice.

Reply:
@dcantah friendly poke

Reply:
Going to try and get to this tonight, a little jammed today 🫠